patches/android_simulcast.patch

```diff
@@ -0,0 +1,193 @@
++    public long createNativeVideoEncoder() {
++        return nativeCreateEncoder(primary, fallback, info);
++    }
++
++    @Override
++    public boolean isHardwareEncoder() {
++        return false;
++    }
++
++}
++
+diff --git a/sdk/android/api/org/webrtc/SimulcastVideoEncoderFactory.java b/sdk/android/api/org/webrtc/SimulcastVideoEncoderFactory.java
+new file mode 100644
+index 0000000000..80f9a50606
+--- /dev/null
++++ b/sdk/android/api/org/webrtc/SimulcastVideoEncoderFactory.java
+@@ -0,0 +1,43 @@
++/*
++ *  Copyright 2017 The WebRTC project authors. All Rights Reserved.
++ *
++ *  Use of this source code is governed by a BSD-style license
++ *  that can be found in the LICENSE file in the root of the source
++ *  tree. An additional intellectual property rights grant can be found
++ *  in the file PATENTS.  All contributing project authors may
++ *  be found in the AUTHORS file in the root of the source tree.
++ */
++
++package org.webrtc;
++
++import android.support.annotation.Nullable;
++import java.util.ArrayList;
++import java.util.HashMap;
++import java.util.List;
++import java.util.Arrays;
++
++public class SimulcastVideoEncoderFactory implements VideoEncoderFactory {
++
++    VideoEncoderFactory primary;
++    VideoEncoderFactory fallback;
++
++    public SimulcastVideoEncoderFactory(VideoEncoderFactory primary, VideoEncoderFactory fallback) {
++        this.primary = primary;
++        this.fallback = fallback;
++    }
++
++    @Nullable
++    @Override
++    public VideoEncoder createEncoder(VideoCodecInfo info) {
++        return new SimulcastVideoEncoder(primary, fallback, info);
++    }
++
++    @Override
++    public VideoCodecInfo[] getSupportedCodecs() {
++        List<VideoCodecInfo> codecs = new ArrayList<VideoCodecInfo>();
++        codecs.addAll(Arrays.asList(primary.getSupportedCodecs()));
++        codecs.addAll(Arrays.asList(fallback.getSupportedCodecs()));
++        return codecs.toArray(new VideoCodecInfo[codecs.size()]);
++    }
++
++}
+
+diff --git a/sdk/android/src/jni/simulcast_video_encoder.cc b/sdk/android/src/jni/simulcast_video_encoder.cc
+new file mode 100644
+index 0000000000..2928cdab92
+--- /dev/null
++++ b/sdk/android/src/jni/simulcast_video_encoder.cc
+@@ -0,0 +1,33 @@
++#include <jni.h>
++
++#include "sdk/android/src/jni/jni_helpers.h"
++#include "sdk/android/src/jni/video_encoder_factory_wrapper.h"
++#include "sdk/android/src/jni/video_codec_info.h"
++#include "sdk/android/native_api/codecs/wrapper.h"
++#include "media/engine/simulcast_encoder_adapter.h"
++#include "rtc_base/logging.h"
++
++using namespace webrtc;
++using namespace webrtc::jni;
++
++#ifdef __cplusplus
++extern "C" {
++#endif
++
++// (VideoEncoderFactory primary, VideoEncoderFactory fallback, VideoCodecInfo info)
++JNIEXPORT jlong JNICALL Java_org_webrtc_SimulcastVideoEncoder_nativeCreateEncoder(JNIEnv *env, jclass klass, jobject primary, jobject fallback, jobject info) {
++    RTC_LOG(LS_INFO) << "Create simulcast video encoder";
```

Comment:
ここは debug レベルでも良いかも

Reply:
あ、この気持ちは、エンコーダはインスタンス化はあまり大事ではなくて、InitEncode が初期化のキモだから、ということです。

Reply:
ありがとうございます
少し調べてみましたが `LS_INFO = デバッグ用` ということで、このままで良さそうです

https://source.chromium.org/chromium/chromium/src/+/master:third_party/webrtc/rtc_base/logging.h;drc=e904161cecbe5e2ca31382e2a62fc776151bb8f2;l=81
> // Note that the non-standard LoggingSeverity aliases exist because they are
// still in broad use.  The meanings of the levels are:
//  LS_VERBOSE: This level is for data which we do not want to appear in the
//   normal debug log, but should appear in diagnostic logs.
//  LS_INFO: Chatty level used in debugging for all sorts of things, the default
//   in debug builds.
//  LS_WARNING: Something that may warrant investigation.
//  LS_ERROR: Something that should not have occurred.
//  LS_NONE: Don't log.

Reply:
👍